### PR TITLE
New JSON representation for JIT compilation

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Instead of serializing the Java representation as JSON, we actually generate a custom JSON representation of a pipeline to match the expectations of the JIT compiler, which is defined as code here and in the related files:
https://github.com/vmware/database-stream-processor/blob/dataflow-jit/crates/dataflow-jit/src/ir

Currently most of the programs would fail to compile to JSON; we have tested Map and Filter operators successfully.
There are many TODOs in the code, which will be filled as I get more understanding of the exact JIT constraints.
But this commit allows the JIT compiler to progress and provides test cases for it.
When that compiler is mature enough we will run side-by-side Rust and JITted code to compare the results.

### CODE REVIEW

File: SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/ToJitVisitor.java [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+                return "I64";
+            case "d":
+                return "F64";
+            case "f":
+                return "F32";
+            case "s":
+                return "String";
+            case "date":
+                return "Date";
+            case "timestamp":
```
Comment (Kixiron) [inline SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/ToJitVisitor.java:138]:
```suggestion
            case "Timestamp":
```

File: SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/ToJitInnerVisitor.java [hunk 2]
```diff
@@ -0,0 +1,706 @@
...
+            binOp.put("operand_ty", baseTypeName(expression.left));
+            if (er.isNullInstruction != null) {
+                // The result is null if either operand is null.
+                binOp = er.isNullInstruction.putObject("BinOp");
+                binOp.put("lhs", leftNullId);
+                binOp.put("rhs", rightNullId);
+                binOp.put("kind", Utilities.getExists(opNames, "||"));
+                binOp.put("operand_ty", "Bool");
+            }
+        }
```
Comment (Kixiron) [inline SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/ToJitInnerVisitor.java:424]:
```suggestion
        ObjectNode binOp = er.instruction.putObject("BinOp");
        binOp.put("lhs", leftId.id);
        binOp.put("rhs", rightId.id);
        binOp.put("kind", Utilities.getExists(opNames, expression.operation));
        binOp.put("operand_ty", baseTypeName(expression.left));
        if (er.isNullInstruction != null) {
            // The result is null if either operand is null.
            binOp = er.isNullInstruction.putObject("BinOp");
            binOp.put("lhs", leftNullId);
            binOp.put("rhs", rightNullId);
            binOp.put("kind", Utilities.getExists(opNames, "||"));
            binOp.put("operand_ty", "Bool");
        }
```

Comment (mbudiu-vmw) [reply SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/ToJitInnerVisitor.java:424]: Unfortunately that's not how SQL works:
https://www.postgresql.org/docs/current/functions-logical.html

Comment (Kixiron) [reply SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/ToJitInnerVisitor.java]: What about it is different? Logical and/or are logical and/or, they're identical from what I can see

Comment (mbudiu-vmw) [reply SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/ToJitInnerVisitor.java]: TRUE || NULL = TRUE instead of NULL

Comment (Kixiron) [reply SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/ToJitInnerVisitor.java]: Can't you just correct the nullish setting stuff then?

**Comment by mbudiu-vmw** [comment]
Rebasing this on main becomes harder and harder as the number of commits grows, so I will resolve the conflicts and merge this branch into main if everything works fine.

**Pull request merged**